{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. First set up amino acid fasta\n",
    "#### 2. Codon optimised and get the tsv file back from IDT (order_optimised_codons.tsv)\n",
    "#### 3. Add any necessary DNA prefix or postfix.\n",
    "#### 4. Get 96-well plate positions file (xlsx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_disulfide_b4inin</td>\n",
       "      <td>TCTCTGGAGGAACTGATCGAAAAAATTGAAAAATTTAACAACCTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no_disulfide_b4innn</td>\n",
       "      <td>CCTCTTGATCCAGAGGCCGTTCGCGCTTGGTTACGCCGTGCTTTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no_disulfide_l5innn</td>\n",
       "      <td>GAGGAGTTGAAAGAGAAACAAGAGGAACTCAAGAAGAAAGTAGACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>no_disulfide_b4nnnnx</td>\n",
       "      <td>TGGGAGGAGTTGGAGGAGGCTGTCCGTTGGCTCCGGGAGCGCCTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>no_disulfide_b4nnnny</td>\n",
       "      <td>ATGAGCCGGGAGGAAAAAATTAAACTTCTGAAGGAGAAGGGCTATA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Name                                           Sequence\n",
       "0   no_disulfide_b4inin  TCTCTGGAGGAACTGATCGAAAAAATTGAAAAATTTAACAACCTCG...\n",
       "1   no_disulfide_b4innn  CCTCTTGATCCAGAGGCCGTTCGCGCTTGGTTACGCCGTGCTTTGG...\n",
       "2   no_disulfide_l5innn  GAGGAGTTGAAAGAGAAACAAGAGGAACTCAAGAAGAAAGTAGACA...\n",
       "3  no_disulfide_b4nnnnx  TGGGAGGAGTTGGAGGAGGCTGTCCGTTGGCTCCGGGAGCGCCTTC...\n",
       "4  no_disulfide_b4nnnny  ATGAGCCGGGAGGAAAAAATTAAACTTCTGAAGGAGAAGGGCTATA..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no_disulfide_b4inin contains disallowed site: CTCGAG\n",
      "no_disulfide_b4inin TCTCTGGAGGAACTGATCGAAAAAATTGAAAAATTTAACAACCTCGAGAATCCTACCGAGGAAGAGCTGGAGGAGCTCATTGAGCTCTTAAAAGAATGGTGTGAGGAACTTGGTGTAGAAGAAGAGAAGAAAGAGGAGCTGTTAGAACTGCTCTACGAGTGGTATAATAGCGAAGAAAATCGCGAAGAGAAGACAAAGGAGCTCATCGAAAAGCTCAAGGAAGTTTACGAGGAGGCG\n",
      "\n",
      "no_disulfide_l4inn contains disallowed site: GGATCC\n",
      "no_disulfide_l4inn GACGAACTGGAAGAAGGGATCCGTTATCTGTACGAGAAGTGGAAGGAAGAGAAGGGGCTGAAATCTCAAGAGGAACTTTTAGAATTACTCCTGGACGAGTCGAATGCTGAAACGGAAGAGGAGAAGGAGATCTATAAGTTGATCAAAGAGTACAAGAAGAAGGGTATGACGGACAAGGATTTATATGAAATTCTGTCGAAAGTTGAAGAGGAG\n",
      "\n",
      "no_disulfide_b6iiniin contains disallowed site: CTCGAG\n",
      "no_disulfide_b6iiniin TCCGAGGAACTCAAAAAATATGCAGAACGGGTAGCGAAGGCGCTTGCCGAGGGGAAGCTCGAGGGTTATTTGAAACTGGTAGAGGAATGGTCCAAAGTAGACAAACCTTTTTCCAAGGAGAACTTCAAAGAGATGCTGGAGAAGGCGCTTTCTACGGAGGAAGCAAAAGAGGATAAAGAACTCAAGGAGGCCCTGGAGTATATCCTCGAAAATCTGGACGAATGGTCCTTAGATGAGATTAAAGAGTACCTGGAATATGCGCTCAAACTGGCGAAGGAATACGGGGTGGAAGAGTACACAAAGTATTTGGAAAAGTTATTATCAAAA\n",
      "\n",
      "no_disulfide_l6innni contains disallowed site: CTCGAG\n",
      "no_disulfide_l6innni TCCCTCTATGAAGAGTTTATGGAATACTTAATGAAAGAAAATGACCTCGAAAAGTTAAAAGAACTTGCGAAAAAGATCGCCGAGAAAATCGGCGCTAAAGATTATATCCCGTTAGTCGAGGAAAGTCTTGAATACCTCAAGGAAGGGAAATGGGAGGAGATGCTTGAGTATATGTTATCGAAGTTGCCGAAGACGAAGCTGCGCGAGGAGTTAGAGGAAATCCTCGAGGGCAAGCTGAGCCCGGAAACAGAGGAGCTGATTAAAGAATGCTATTCA\n",
      "\n",
      "variable_linkers_l5innn contains disallowed site: CTCGAG\n",
      "variable_linkers_l5innn ATGAAGAAGAAGGTCGCAAAGCAGCTCGAGGAAGGCAAGGTCGACCCGGAATTACTTGCCAAGGTCTATAAAGAGCTGACAGGCGATGAGTCGACGGCCAAAATTGTCGCTAAAGTGGCAAAAAAGAAATTGAGTAAGGAGGAATTAGCTAAGGAAATGAAAAAGGAAGGTGTGACCATCTCCGCTAAAGCTCTCGAGAAGTTA\n",
      "\n",
      "variable_linkers_l5inni contains disallowed site: CTCGAG\n",
      "variable_linkers_l5inni ATGAAAGAAAAATACGCGAAACTCCTCGAGGAGGTGGCTAAAGAGACCAATGAGGTTATGAAGGAAATTGCCGAAATGCTGAAGAACGTGGACGAAGAAAACCCTCTGGAGGATCTCAAAAAGAAGATCATGGAGAAGTTCGGCGCATCGGAAGAGGTTGCCGAAGAGTTTGCAAAGGCCCTTAAGGCTGGTGACAAAGAAAAGTTGAAGAAATGCGTTAAAAAGATCCTTGAA\n",
      "\n",
      "variable_linkers_b3iii contains disallowed site: CTCGAG\n",
      "variable_linkers_b3iii ATGGAAGAGAAAGCGAAAGAGATCGTTGAACTGTATAAGGCGGGGAAGCTCACGGAGGAGCAGAAGAAAATTGCGAAGGAAATCCTCGAGATCGAGAAAAAGGAAAAGAAATTTTCTGAGGAGTTGGCTAAGAAGGTGGAAGAAGTCGTAAAA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "opt_codons_df = pd.read_csv(\"/home/tadas/code/single_chain_dp_bristol/order_optimised_codons.tsv\", sep=\"\\t\",header=None)\n",
    "opt_codons_df.rename(columns={0:\"Name\",1:\"Sequence\"}, inplace=True)\n",
    "assert all(len(x) % 3 == 0 for x in opt_codons_df[\"Sequence\"])\n",
    "display(opt_codons_df.head())\n",
    "disalowed_sites = [\"CTCGAG\",\"GGATCC\"]\n",
    "\n",
    "def report_disallowed_sites(opt_codons_df, disalowed_sites,tolerance = 0):\n",
    "    for name, seq in zip(opt_codons_df[\"Name\"], opt_codons_df[\"Sequence\"]):\n",
    "        for site in disalowed_sites:\n",
    "            # alow number of sites = tolerance\n",
    "            num_sites = seq.count(site)\n",
    "            if num_sites > tolerance:\n",
    "                print(f\"{name} contains disallowed site: {site}\")\n",
    "                print(name, seq)\n",
    "                print()\n",
    "        # alowed_dna =  [site not in x for x in opt_codons_df[\"Sequence\"]]\n",
    "        # print(f\"{sum(alowed_dna)}/{len(opt_codons_df)} sequences do not contain disallowed site: {site}\")\n",
    "        # assert all(alowed_dna), f\"{sum(alowed_dna)}/{len(opt_codons_df)} sequences do not contain disallowed site: {site}\"\n",
    "\n",
    "report_disallowed_sites(opt_codons_df, disalowed_sites)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove the disallowed cut sites using Benchling! \n",
    "##### then export it back to order_optimised_codons_no_cuts.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "opt_codons_no_cuts_df = pd.read_csv(\"/home/tadas/code/single_chain_dp_bristol/order_optimised_codons_no_cuts.csv\")\n",
    "opt_codons_no_cuts_df.head()\n",
    "assert all(len(x) % 3 == 0 for x in opt_codons_df[\"Sequence\"])\n",
    "report_disallowed_sites(opt_codons_no_cuts_df, disalowed_sites)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make sure there are no cut sites in this file\n",
    "#### Remove any unnecessary names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Author</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>no_disulfide_b3iii</td>\n",
       "      <td>Tadas</td>\n",
       "      <td>ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>no_disulfide_b3nnn</td>\n",
       "      <td>Tadas</td>\n",
       "      <td>ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no_disulfide_b4iiiix</td>\n",
       "      <td>Tadas</td>\n",
       "      <td>ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>no_disulfide_b4iiiiy</td>\n",
       "      <td>Tadas</td>\n",
       "      <td>ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>no_disulfide_b4iiin</td>\n",
       "      <td>Tadas</td>\n",
       "      <td>ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Name Author  \\\n",
       "0    no_disulfide_b3iii  Tadas   \n",
       "1    no_disulfide_b3nnn  Tadas   \n",
       "2  no_disulfide_b4iiiix  Tadas   \n",
       "3  no_disulfide_b4iiiiy  Tadas   \n",
       "4   no_disulfide_b4iiin  Tadas   \n",
       "\n",
       "                                            Sequence  \n",
       "0  ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...  \n",
       "1  ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...  \n",
       "2  ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...  \n",
       "3  ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...  \n",
       "4  ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "prefix = \"ATGGGCAGCAGCCATCATCATCATCATCACAGCAGCGGCGAGAACTTATACTTTCAGAGCGGATCC\"\n",
    "assert len(prefix) % 3 == 0\n",
    "# MGSSHHHHHHSSGENLYFQSGS\n",
    "# Start codon, His-tag, flexible linkers, TEV cut site, BamHI site\n",
    "postfix = \"TAACTCGAG\"\n",
    "assert postfix.startswith(\"TAA\")\n",
    "# Stop codon, XhoI site\n",
    "# add perfix and postfix to opt_codons_no_cuts_df Sequence column values\n",
    "opt_codons_no_cuts_df[\"Sequence\"] = opt_codons_no_cuts_df[\"Sequence\"].apply(lambda x: prefix+x+postfix)\n",
    "# check that all Sequence lengths are multiples of 3\n",
    "\n",
    "display(opt_codons_no_cuts_df.head())\n",
    "# assert that all sequences are <500 bp \n",
    "assert all(len(x) >= 125 for x in opt_codons_no_cuts_df[\"Sequence\"])\n",
    "assert all(len(x) <= 500 for x in opt_codons_no_cuts_df[\"Sequence\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now assign 96-well plate positions for IDT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "def assign_well_positions(df, manual_breaks):\n",
    "    \"\"\"\n",
    "    Assign well positions on a 96-well plate to the DataFrame.\n",
    "    \n",
    "    The plate is assumed to have 8 rows (A-H) and 12 columns. The DataFrame's\n",
    "    rows are filled in order from left to right. If a row index is specified in\n",
    "    `manual_breaks`, that sample starts at column 1 of a new row.\n",
    "    \n",
    "    Parameters:\n",
    "        df (pd.DataFrame): DataFrame with at least columns \"Name\" and \"Sequence\".\n",
    "        manual_breaks (list of int): List of zero-indexed row numbers that should\n",
    "                                     start a new row in the plate (e.g. [15, 21]).\n",
    "                                     \n",
    "    Returns:\n",
    "        pd.DataFrame: The original DataFrame with an added \"Well Position\" column.\n",
    "    \"\"\"\n",
    "    # Reset index to ensure sequential numbering\n",
    "    df = df.reset_index(drop=True)\n",
    "    \n",
    "    # Define plate rows and the maximum number of columns\n",
    "    plate_rows = list(\"ABCDEFGH\")\n",
    "    max_col = 12\n",
    "    \n",
    "    well_positions = []\n",
    "    current_row = 0  # index for plate_rows (starting with \"A\")\n",
    "    current_col = 1  # column number, from 1 to 12\n",
    "    \n",
    "    for i in range(len(df)):\n",
    "        # If this row index is in manual_breaks (and it's not the very first sample),\n",
    "        # force a new row by incrementing the row counter and resetting the column.\n",
    "        if i in manual_breaks and i != 0:\n",
    "            if not current_col == 1:\n",
    "                current_row += 1\n",
    "                current_col = 1\n",
    "\n",
    "        # Automatically start a new row if the current column exceeds max_col\n",
    "        if current_col > max_col:\n",
    "            current_row += 1\n",
    "            current_col = 1\n",
    "        \n",
    "        # Ensure we haven't exceeded the available plate rows\n",
    "        if current_row >= len(plate_rows):\n",
    "            raise ValueError(\"Not enough plate rows available for the given samples.\")\n",
    "        \n",
    "        # Create the well position string, e.g., \"A1\", \"B5\", etc.\n",
    "        well = f\"{plate_rows[current_row]}{current_col}\"\n",
    "        well_positions.append(well)\n",
    "        \n",
    "        # Move to the next column for the next sample\n",
    "        current_col += 1\n",
    "    \n",
    "    # Add the \"Well Position\" column to the DataFrame and return it\n",
    "    df[\"Well Position\"] = well_positions\n",
    "    return df[[\"Well Position\", \"Name\", \"Sequence\"]]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "break_after_entries = [30]\n",
    "\n",
    "opt_codons_no_cuts_df = assign_well_positions(opt_codons_no_cuts_df, break_after_entries)\n",
    "report_disallowed_sites(opt_codons_no_cuts_df, disalowed_sites, tolerance = 1)\n",
    "opt_codons_no_cuts_df.to_excel(\"/home/tadas/code/single_chain_dp_bristol/order_optimised_codons_96_wp.xlsx\", index=False)\n",
    "opt_codons_no_cuts_df.to_csv(\"/home/tadas/code/single_chain_dp_bristol/order_optimised_codons_96_wp.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
